main.tf

```
@@ -17,126 +21,194 @@ resource "aws_s3_bucket" "default" {
   #bridgecrew:skip=BC_AWS_S3_16:Skipping `Ensure S3 bucket versioning is enabled` because dynamic blocks are not supported by checkov
   #bridgecrew:skip=BC_AWS_S3_14:Skipping `Ensure all data stored in the S3 bucket is securely encrypted at rest` because variables are not understood
   #bridgecrew:skip=BC_AWS_GENERAL_56:Skipping `Ensure that S3 buckets are encrypted with KMS by default` because we do not have good defaults
-  count               = local.enabled ? 1 : 0
-  bucket              = local.bucket_name
-  acl                 = try(length(var.grants), 0) == 0 ? var.acl : null
-  force_destroy       = var.force_destroy
-  tags                = module.this.tags
-  acceleration_status = var.transfer_acceleration_enabled ? "Enabled" : null
-
-  dynamic "versioning" {
-    for_each = var.versioning_enabled ? [true] : []
+  count         = local.enabled ? 1 : 0
+  bucket        = local.bucket_name
+  force_destroy = var.force_destroy
+
+  dynamic "object_lock_configuration" {
+    for_each = var.object_lock_configuration != null ? [1] : []
+
     content {
-      enabled = true
+      object_lock_enabled = "Enabled"
     }
   }
 
-  dynamic "lifecycle_rule" {
+  tags = module.this.tags
+}
+
+resource "aws_s3_bucket_accelerate_configuration" "default" {
+  count  = local.transfer_acceleration_enabled ? 1 : 0
+  bucket = join("", aws_s3_bucket.default.*.id)
+  status = "Enabled"
+}
+
+resource "aws_s3_bucket_versioning" "default" {
+  count = local.versioning_enabled ? 1 : 0
+
+  bucket = join("", aws_s3_bucket.default.*.id)
+
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
+
+resource "aws_s3_bucket_lifecycle_configuration" "default" {
+  count  = local.enabled && length(var.lifecycle_rules) > 0 ? 1 : 0
+  bucket = join("", aws_s3_bucket.default.*.id)
+
+  dynamic "rule" {
     for_each = var.lifecycle_rules
+
     content {
-      enabled                                = lifecycle_rule.value.enabled
-      prefix                                 = lifecycle_rule.value.prefix
-      tags                                   = lifecycle_rule.value.tags
-      abort_incomplete_multipart_upload_days = lifecycle_rule.value.abort_incomplete_multipart_upload_days
+      id     = rule.value.id
+      status = try(rule.value.enabled == true ? "Enabled" : "Disabled", rule.value.status)
```

We don't have the `status` attribute in our `var.lifecycle_rules` objects, so we cannot fall back to it.

Also, since the variable is `list(object{...})` and not `list(any)` and hence the objects are actually typed, we don't need the `try()` function.

```suggestion
      status = rule.value.enabled == true ? "Enabled" : "Disabled"
```

---

This was a backwards compatibility attempt